File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
         if(head == NULL && head->next ==NULL) return;
        ListNode* mid = getMid(head);
        ListNode* curr = mid;
        ListNode* prev = NULL;
        ListNode* next = NULL;
        while(curr != NULL){
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        ListNode* h1 = head;
        ListNode* h2 = prev;
        next = NULL;
        while(h1 != h2){
            next = h1->next;
            h1->next = h2;
            h1 = next;

            next = h2->next;
            if(next == NULL) break;
            h2->next = h1;
            h2 = next;
        }
    }

    ListNode* getMid(ListNode* head){
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast != NULL && fast->next != NULL){
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
};

File: 0143-reorder-list/0143-reorder-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public void ReorderList(ListNode head) {
        if(head == null && head.next ==null) return;
        
        //Getting Middle Node
        ListNode mid = head;
        ListNode fast = head;
        while(fast != null && fast.next != null){
            mid = mid.next;
            fast = fast.next.next;
        }
        
        //Reverse the list from Middle
        ListNode prev = null;
        ListNode next = null;
        while(mid != null){
            next = mid.next;
            mid.next = prev;
            prev = mid;
            mid = next;
        }
        ListNode h1 = head;
        ListNode h2 = prev;
        next = null;
        while(h1 != h2){
            next = h1.next;
            h1.next = h2;
            h1 = next;
            next = h2.next;
            if(next == null) break;
            h2.next = h1;
            h2 = next;
        }
    }
}